File: python/RAG-MODULE/image-processing/cleanup_image_collections.py
#!/usr/bin/env python3.9
"""
Cleanup Script for Image Collections
Deletes user image collections from filesystem and provides ChromaDB cleanup info
"""

import os
import sys
import json
import shutil
import argparse
from pathlib import Path
from datetime import datetime

class ImageCollectionCleaner:
    def __init__(self, base_data_dir="/app/data"):
        self.base_data_dir = Path(base_data_dir)
        self.collections_dir = self.base_data_dir / "collections"
    
    def list_image_collections(self):
        """List all user image collections"""
        collections = []
        
        if not self.collections_dir.exists():
            print("📁 No collections directory found")
            return collections
        
        try:
            for entry in self.collections_dir.iterdir():
                if entry.is_dir() and "_images" in entry.name:
                    info = self.get_collection_info(entry)
                    collections.append({
                        "name": entry.name,
                        "path": str(entry),
                        **info
                    })
            
            return collections
        except Exception as e:
            print(f"❌ Error listing collections: {e}")
            return []
    
    def get_collection_info(self, collection_path):
        """Get information about a collection"""
        try:
            sessions = []
            total_images = 0
            total_size = 0
            
            for entry in collection_path.iterdir():
                if entry.is_dir():
                    session_info = self.get_session_info(entry)
                    sessions.append({
                        "name": entry.name,
                        "path": str(entry),
                        **session_info
                    })
                    total_images += session_info["image_count"]
                    total_size += session_info["total_size"]
            
            return {
                "sessions": sessions,
                "total_images": total_images,
                "total_size_kb": round(total_size / 1024),
                "last_modified": collection_path.stat().st_mtime
            }
        except Exception as e:
            return {
                "sessions": [],
                "total_images": 0,
                "total_size_kb": 0,
                "error": str(e)
            }
    
    def get_session_info(self, session_path):
        """Get information about a session"""
        try:
            image_count = 0
            total_size = 0
            metadata_file = None
            
            for entry in session_path.iterdir():
                if entry.name == "collection_metadata.json":
                    metadata_file = entry
                elif entry.name.startswith("img_"):
                    image_count += 1
                    total_size += entry.stat().st_size
            
            metadata = None
            if metadata_file and metadata_file.exists():
                try:
                    with open(metadata_file, 'r') as f:
                        metadata = json.load(f)
                except Exception as e:
                    print(f"⚠️ Could not read metadata: {e}")
            
            return {
                "image_count": image_count,
                "total_size": total_size,
                "metadata": metadata,
                "last_modified": session_path.stat().st_mtime
            }
        except Exception as e:
            return {
                "image_count": 0,
                "total_size": 0,
                "metadata": None,
                "error": str(e)
            }
    
    def delete_user_collection(self, user_id, session_id=None):
        """Delete a specific user's image collection"""
        try:
            print(f"🗑️ Deleting collection for user: {user_id}{f', session: {session_id}' if session_id else ''}")
            
            user_collection_name = f"user_{user_id.replace('-', '_')}_images"
            user_collection_path = self.collections_dir / user_collection_name
            
            if not user_collection_path.exists():
                print(f"📁 Collection not found: {user_collection_name}")
                return {"success": False, "message": "Collection not found"}
            
            if session_id:
                # Delete specific session
                session_name = f"session_{session_id.replace('-', '_')}"
                session_path = user_collection_path / session_name
                
                if session_path.exists():
                    shutil.rmtree(session_path)
                    print(f"✅ Deleted session: {session_name}")
                    
                    # Check if user collection is now empty
                    remaining_sessions = list(user_collection_path.iterdir())
                    if not remaining_sessions:
                        shutil.rmtree(user_collection_path)
                        print(f"✅ Deleted empty user collection: {user_collection_name}")
                else:
                    print(f"📁 Session not found: {session_name}")
                    return {"success": False, "message": "Session not found"}
            else:
                # Delete entire user collection
                shutil.rmtree(user_collection_path)
                print(f"✅ Deleted user collection: {user_collection_name}")
            
            return {"success": True, "message": "Collection deleted successfully"}
        except Exception as e:
            print(f"❌ Error deleting collection: {e}")
            return {"success": False, "error": str(e)}
    
    def delete_all_image_collections(self):
        """Delete all image collections"""
        try:
            print("🗑️ Deleting ALL image collections...")
            
            collections = self.list_image_collections()
            
            if not collections:
                print("📁 No image collections found")
                return {"success": True, "message": "No collections to delete"}
            
            deleted_count = 0
            for collection in collections:
                try:
                    shutil.rmtree(collection["path"])
                    print(f"✅ Deleted: {collection['name']}")
                    deleted_count += 1
                except Exception as e:
                    print(f"❌ Failed to delete {collection['name']}: {e}")
            
            print(f"🎯 Deleted {deleted_count} out of {len(collections)} collections")
            return {"success": True, "deleted_count": deleted_count, "total_count": len(collections)}
        except Exception as e:
            print(f"❌ Error deleting all collections: {e}")
            return {"success": False, "error": str(e)}
    
    def display_collections_summary(self):
        """Display collections summary"""
        print("📊 IMAGE COLLECTIONS SUMMARY")
        print("=" * 40)
        
        collections = self.list_image_collections()
        
        if not collections:
            print("📁 No image collections found")
            return
        
        total_images = 0
        total_size = 0
        
        for collection in collections:
            print(f"\n👤 {collection['name']}")
            print(f"   📁 Path: {collection['path']}")
            print(f"   🖼️ Images: {collection['total_images']}")
            print(f"   💾 Size: {collection['total_size_kb']}KB")
            
            # Convert timestamp to readable date
            last_modified = datetime.fromtimestamp(collection['last_modified'])
            print(f"   📅 Modified: {last_modified.strftime('%Y-%m-%d %H:%M:%S')}")
            
            if collection['sessions']:
                print("   📂 Sessions:")
                for session in collection['sessions']:
                    size_kb = round(session['total_size'] / 1024)
                    print(f"      - {session['name']}: {session['image_count']} images ({size_kb}KB)")
                    
                    # Show metadata if available
                    if session['metadata']:
                        meta = session['metadata']
                        if 'pdf_source' in meta:
                            print(f"        📄 Source: {meta['pdf_source']}")
                        if 'created_at' in meta:
                            print(f"        📅 Created: {meta['created_at']}")
            
            total_images += collection['total_images']
            total_size += collection['total_size_kb']
        
        print(f"\n📈 TOTALS:")
        print(f"   Collections: {len(collections)}")
        print(f"   Total Images: {total_images}")
        print(f"   Total Size: {total_size}KB")
    
    def show_keywords_summary(self):
        """Show keywords from all collections"""
        print("🔤 KEYWORDS SUMMARY FROM ALL COLLECTIONS")
        print("=" * 50)
        
        collections = self.list_image_collections()
        all_keywords = []
        
        for collection in collections:
            for session in collection['sessions']:
                if session['metadata'] and 'images' in session['metadata']:
                    for img in session['metadata']['images']:
                        keywords = img.get('keywords', '')
                        if keywords and 'Page ' not in keywords:  # Skip fallback keywords
                            all_keywords.append({
                                'user': collection['name'],
                                'session': session['name'],
                                'page': img.get('page', 'N/A'),
                                'filename': img.get('filename', 'N/A'),
                                'keywords': keywords
                            })
        
        if not all_keywords:
            print("📁 No keywords found in collections")
            return
        
        print(f"Found {len(all_keywords)} images with OCR keywords:")
        print()
        
        for i, item in enumerate(all_keywords[:20], 1):  # Show first 20
            print(f"{i:2d}. {item['user']} | {item['session']} | Page {item['page']} | {item['keywords'][:80]}...")
        
        if len(all_keywords) > 20:
            print(f"\n... and {len(all_keywords) - 20} more images with keywords")


def main():
    """Main CLI function"""
    parser = argparse.ArgumentParser(description="Image Collection Cleanup Tool")
    parser.add_argument("command", choices=["list", "delete-user", "delete-session", "delete-all", "keywords"], 
                       help="Command to execute")
    parser.add_argument("--user-id", help="User ID for delete-user and delete-session commands")
    parser.add_argument("--session-id", help="Session ID for delete-session command")
    parser.add_argument("--data-dir", default="/app/data", help="Base data directory")
    
    if len(sys.argv) == 1:
        print("🧹 Image Collection Cleanup Tool")
        print("================================")
        print("Usage:")
        print("  python cleanup_image_collections.py list                           # List all collections")
        print("  python cleanup_image_collections.py delete-user --user-id <userId> # Delete user collection")
        print("  python cleanup_image_collections.py delete-session --user-id <userId> --session-id <sessionId>")
        print("  python cleanup_image_collections.py delete-all                     # Delete all collections")
        print("  python cleanup_image_collections.py keywords                       # Show all extracted keywords")
        return
    
    args = parser.parse_args()
    cleaner = ImageCollectionCleaner(args.data_dir)
    
    if args.command == "list":
        cleaner.display_collections_summary()
    
    elif args.command == "delete-user":
        if not args.user_id:
            print("❌ --user-id is required for delete-user command")
            sys.exit(1)
        result = cleaner.delete_user_collection(args.user_id)
        print("✅ Success" if result["success"] else "❌ Failed:", result.get("message", result.get("error")))
    
    elif args.command == "delete-session":
        if not args.user_id or not args.session_id:
            print("❌ Both --user-id and --session-id are required for delete-session command")
            sys.exit(1)
        result = cleaner.delete_user_collection(args.user_id, args.session_id)
        print("✅ Success" if result["success"] else "❌ Failed:", result.get("message", result.get("error")))
    
    elif args.command == "delete-all":
        result = cleaner.delete_all_image_collections()
        print("✅ Success" if result["success"] else "❌ Failed:", result.get("message", result.get("error")))
    
    elif args.command == "keywords":
        cleaner.show_keywords_summary()


if __name__ == "__main__":
    main()

File: python/RAG-MODULE/image-processing/cleanup_image_collections.js
#!/usr/bin/env node
/**
 * Cleanup Script for Image Collections
 * Deletes user image collections from both filesystem and ChromaDB
 */

const fs = require('fs').promises;
const path = require('path');
const { spawn } = require('child_process');

class ImageCollectionCleaner {
  constructor() {
    this.baseDataDir = '/app/data';
    this.collectionsDir = path.join(this.baseDataDir, 'collections');
    this.chromaDbHost = process.env.CHROMADB_HOST || 'chromadb';
    this.chromaDbPort = process.env.CHROMADB_PORT || '8000';
  }

  /**
   * List all user image collections
   */
  async listImageCollections() {
    try {
      const collections = [];
      
      if (!(await this.directoryExists(this.collectionsDir))) {
        console.log('📁 No collections directory found');
        return collections;
      }

      const entries = await fs.readdir(this.collectionsDir);
      
      for (const entry of entries) {
        if (entry.includes('_images')) {
          const collectionPath = path.join(this.collectionsDir, entry);
          const stats = await fs.stat(collectionPath);
          
          if (stats.isDirectory()) {
            const info = await this.getCollectionInfo(collectionPath);
            collections.push({
              name: entry,
              path: collectionPath,
              ...info
            });
          }
        }
      }

      return collections;
    } catch (error) {
      console.error('❌ Error listing collections:', error.message);
      return [];
    }
  }

  /**
   * Get information about a collection
   */
  async getCollectionInfo(collectionPath) {
    try {
      const sessions = [];
      let totalImages = 0;
      let totalSize = 0;

      const entries = await fs.readdir(collectionPath);
      
      for (const entry of entries) {
        const sessionPath = path.join(collectionPath, entry);
        const stats = await fs.stat(sessionPath);
        
        if (stats.isDirectory()) {
          const sessionInfo = await this.getSessionInfo(sessionPath);
          sessions.push({
            name: entry,
            path: sessionPath,
            ...sessionInfo
          });
          totalImages += sessionInfo.imageCount;
          totalSize += sessionInfo.totalSize;
        }
      }

      return {
        sessions,
        totalImages,
        totalSize: Math.round(totalSize / 1024), // Convert to KB
        lastModified: (await fs.stat(collectionPath)).mtime
      };
    } catch (error) {
      return {
        sessions: [],
        totalImages: 0,
        totalSize: 0,
        error: error.message
      };
    }
  }

  /**
   * Get information about a session
   */
  async getSessionInfo(sessionPath) {
    try {
      const entries = await fs.readdir(sessionPath);
      let imageCount = 0;
      let totalSize = 0;
      let metadataFile = null;

      for (const entry of entries) {
        const filePath = path.join(sessionPath, entry);
        const stats = await fs.stat(filePath);
        
        if (entry === 'collection_metadata.json') {
          metadataFile = filePath;
        } else if (entry.startsWith('img_')) {
          imageCount++;
          totalSize += stats.size;
        }
      }

      let metadata = null;
      if (metadataFile) {
        try {
          const content = await fs.readFile(metadataFile, 'utf8');
          metadata = JSON.parse(content);
        } catch (e) {
          console.warn(`⚠️ Could not read metadata: ${e.message}`);
        }
      }

      return {
        imageCount,
        totalSize,
        metadata,
        lastModified: (await fs.stat(sessionPath)).mtime
      };
    } catch (error) {
      return {
        imageCount: 0,
        totalSize: 0,
        metadata: null,
        error: error.message
      };
    }
  }

  /**
   * Delete a specific user's image collection
   */
  async deleteUserCollection(userId, sessionId = null) {
    try {
      console.log(`🗑️ Deleting collection for user: ${userId}${sessionId ? `, session: ${sessionId}` : ''}`);
      
      const userCollectionName = `user_${userId.replace(/-/g, '_')}_images`;
      const userCollectionPath = path.join(this.collectionsDir, userCollectionName);

      if (!(await this.directoryExists(userCollectionPath))) {
        console.log(`📁 Collection not found: ${userCollectionName}`);
        return { success: false, message: 'Collection not found' };
      }

      if (sessionId) {
        // Delete specific session
        const sessionName = `session_${sessionId.replace(/-/g, '_')}`;
        const sessionPath = path.join(userCollectionPath, sessionName);
        
        if (await this.directoryExists(sessionPath)) {
          await this.deleteDirectory(sessionPath);
          console.log(`✅ Deleted session: ${sessionName}`);
          
          // Check if user collection is now empty
          const remainingSessions = await fs.readdir(userCollectionPath);
          if (remainingSessions.length === 0) {
            await this.deleteDirectory(userCollectionPath);
            console.log(`✅ Deleted empty user collection: ${userCollectionName}`);
          }
        } else {
          console.log(`📁 Session not found: ${sessionName}`);
          return { success: false, message: 'Session not found' };
        }
      } else {
        // Delete entire user collection
        await this.deleteDirectory(userCollectionPath);
        console.log(`✅ Deleted user collection: ${userCollectionName}`);
      }

      return { success: true, message: 'Collection deleted successfully' };
    } catch (error) {
      console.error(`❌ Error deleting collection: ${error.message}`);
      return { success: false, error: error.message };
    }
  }

  /**
   * Delete all image collections
   */
  async deleteAllImageCollections() {
    try {
      console.log('🗑️ Deleting ALL image collections...');
      
      const collections = await this.listImageCollections();
      
      if (collections.length === 0) {
        console.log('📁 No image collections found');
        return { success: true, message: 'No collections to delete' };
      }

      let deletedCount = 0;
      for (const collection of collections) {
        try {
          await this.deleteDirectory(collection.path);
          console.log(`✅ Deleted: ${collection.name}`);
          deletedCount++;
        } catch (error) {
          console.error(`❌ Failed to delete ${collection.name}: ${error.message}`);
        }
      }

      console.log(`🎯 Deleted ${deletedCount} out of ${collections.length} collections`);
      return { success: true, deletedCount, totalCount: collections.length };
    } catch (error) {
      console.error(`❌ Error deleting all collections: ${error.message}`);
      return { success: false, error: error.message };
    }
  }

  /**
   * Clean up ChromaDB image collections (if they exist)
   */
  async cleanupChromaDbImageCollections() {
    try {
      console.log('🗄️ Cleaning up ChromaDB image collections...');
      
      // This would require ChromaDB client integration
      // For now, just log the action
      console.log('ℹ️ ChromaDB cleanup would be implemented here');
      console.log('   - List all collections with "_images" suffix');
      console.log('   - Delete collections that match image collection pattern');
      
      return { success: true, message: 'ChromaDB cleanup completed' };
    } catch (error) {
      console.error(`❌ ChromaDB cleanup error: ${error.message}`);
      return { success: false, error: error.message };
    }
  }

  /**
   * Utility: Check if directory exists
   */
  async directoryExists(dirPath) {
    try {
      const stats = await fs.stat(dirPath);
      return stats.isDirectory();
    } catch {
      return false;
    }
  }

  /**
   * Utility: Delete directory recursively
   */
  async deleteDirectory(dirPath) {
    await fs.rm(dirPath, { recursive: true, force: true });
  }

  /**
   * Display collections summary
   */
  async displayCollectionsSummary() {
    console.log('📊 IMAGE COLLECTIONS SUMMARY');
    console.log('=' * 40);
    
    const collections = await this.listImageCollections();
    
    if (collections.length === 0) {
      console.log('📁 No image collections found');
      return;
    }

    let totalImages = 0;
    let totalSize = 0;

    for (const collection of collections) {
      console.log(`\n👤 ${collection.name}`);
      console.log(`   📁 Path: ${collection.path}`);
      console.log(`   🖼️ Images: ${collection.totalImages}`);
      console.log(`   💾 Size: ${collection.totalSize}KB`);
      console.log(`   📅 Modified: ${collection.lastModified.toISOString()}`);
      
      if (collection.sessions.length > 0) {
        console.log(`   📂 Sessions:`);
        for (const session of collection.sessions) {
          console.log(`      - ${session.name}: ${session.imageCount} images (${Math.round(session.totalSize/1024)}KB)`);
        }
      }

      totalImages += collection.totalImages;
      totalSize += collection.totalSize;
    }

    console.log(`\n📈 TOTALS:`);
    console.log(`   Collections: ${collections.length}`);
    console.log(`   Total Images: ${totalImages}`);
    console.log(`   Total Size: ${totalSize}KB`);
  }
}

/**
 * Main CLI function
 */
async function main() {
  const args = process.argv.slice(2);
  const cleaner = new ImageCollectionCleaner();

  if (args.length === 0) {
    console.log('🧹 Image Collection Cleanup Tool');
    console.log('================================');
    console.log('Usage:');
    console.log('  node cleanup_image_collections.js list                    # List all collections');
    console.log('  node cleanup_image_collections.js delete-user <userId>    # Delete user collection');
    console.log('  node cleanup_image_collections.js delete-session <userId> <sessionId>  # Delete specific session');
    console.log('  node cleanup_image_collections.js delete-all              # Delete all collections');
    console.log('  node cleanup_image_collections.js cleanup-chromadb        # Cleanup ChromaDB');
    return;
  }

  const command = args[0];

  switch (command) {
    case 'list':
      await cleaner.displayCollectionsSummary();
      break;

    case 'delete-user':
      if (args.length < 2) {
        console.error('❌ Usage: delete-user <userId>');
        process.exit(1);
      }
      const result = await cleaner.deleteUserCollection(args[1]);
      console.log(result.success ? '✅ Success' : '❌ Failed:', result.message || result.error);
      break;

    case 'delete-session':
      if (args.length < 3) {
        console.error('❌ Usage: delete-session <userId> <sessionId>');
        process.exit(1);
      }
      const sessionResult = await cleaner.deleteUserCollection(args[1], args[2]);
      console.log(sessionResult.success ? '✅ Success' : '❌ Failed:', sessionResult.message || sessionResult.error);
      break;

    case 'delete-all':
      const allResult = await cleaner.deleteAllImageCollections();
      console.log(allResult.success ? '✅ Success' : '❌ Failed:', allResult.message || allResult.error);
      break;

    case 'cleanup-chromadb':
      const chromaResult = await cleaner.cleanupChromaDbImageCollections();
      console.log(chromaResult.success ? '✅ Success' : '❌ Failed:', chromaResult.message || chromaResult.error);
      break;

    default:
      console.error(`❌ Unknown command: ${command}`);
      process.exit(1);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = ImageCollectionCleaner;
